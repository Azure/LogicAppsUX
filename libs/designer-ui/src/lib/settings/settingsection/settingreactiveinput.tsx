import { SettingTextField } from './settingtextfield';
import type { TextInputChangeHandler } from './settingtextfield';
import { SettingToggle } from './settingtoggle';
import type { SettingProps } from './settingtoggle';
import { useState } from 'react';

export interface ReactiveToggleProps extends SettingProps {
  textFieldLabel: string;
  textFieldId?: string;
  textFieldValue: string;
  checked?: boolean;
  onToggleLabel: string;
  offToggleLabel: string;
  onValueChange?: TextInputChangeHandler;
}

export const ReactiveToggle: React.FC<ReactiveToggleProps> = ({
  textFieldValue,
  textFieldLabel,
  textFieldId,
  readOnly = false,
  checked,
  customLabel,
  onToggleLabel,
  offToggleLabel,
  onValueChange,
  ariaLabel,
}: ReactiveToggleProps): JSX.Element | null => {
  const [checkedState, setChecked] = useState(checked ?? false);
  const onToggleInputChange = (e: React.MouseEvent<HTMLElement>, checked?: boolean): void => {
    e.stopPropagation();
    e.preventDefault();
    setChecked(!!checked);
  };

  if (customLabel) {
    return (
      <>
        {customLabel()}
        <SettingToggle
          checked={checkedState}
          ariaLabel={ariaLabel}
          readOnly={readOnly}
          onToggleInputChange={onToggleInputChange}
          onText={onToggleLabel}
          offText={offToggleLabel}
        />
        {checkedState ? (
          <SettingTextField
            id={textFieldId}
            ariaLabel={ariaLabel}
            value={textFieldValue}
            label={textFieldLabel}
            readOnly={readOnly}
            onValueChange={onValueChange}
          />
        ) : null}
      </>
    );
  } else {
    return (
      <>
        <SettingToggle
          checked={checkedState}
          readOnly={readOnly}
          ariaLabel={ariaLabel}
          onToggleInputChange={onToggleInputChange}
          onText={onToggleLabel}
          offText={offToggleLabel}
        />
        {checkedState ? (
          <SettingTextField
            id={textFieldId}
            ariaLabel={ariaLabel}
            value={textFieldValue}
            label={textFieldLabel}
            readOnly={readOnly}
            onValueChange={onValueChange}
          />
        ) : null}
      </>
    );
  }
};
