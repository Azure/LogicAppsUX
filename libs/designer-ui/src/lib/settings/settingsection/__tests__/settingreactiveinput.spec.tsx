import type { ReactiveToggleProps } from '../settingreactiveinput';
import { ReactiveToggle } from '../settingreactiveinput';
import React from 'react';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';

describe('ui/settings/settingreactiveinput', () => {
  let minimal: ReactiveToggleProps, renderer: ReactShallowRenderer.ShallowRenderer;

  beforeEach(() => {
    minimal = {
      textFieldId: 'id1',
      readOnly: false,
      textFieldLabel: 'text field label',
      textFieldValue: 'test',
      defaultChecked: false,
    };
    renderer = ReactShallowRenderer.createRenderer();
  });

  afterEach(() => {
    renderer.unmount();
  });

  it('should construct', () => {
    const settingToggle = renderer.render(<ReactiveToggle {...minimal} />);
    expect(settingToggle).toMatchSnapshot();
  });

  it('should have child text field with correct value', () => {
    const props: ReactiveToggleProps = { ...minimal, defaultChecked: true };
    renderer.render(<ReactiveToggle {...props} />);

    const reactiveToggle = renderer.getRenderOutput();
    const [toggle, textField]: any[] = React.Children.toArray(reactiveToggle.props.children);

    expect(toggle.props.checked).toBeTruthy();
    expect(textField.props.id).toBe(minimal.textFieldId);
    expect(textField.props.value).toBe(minimal.textFieldValue);
  });
});
