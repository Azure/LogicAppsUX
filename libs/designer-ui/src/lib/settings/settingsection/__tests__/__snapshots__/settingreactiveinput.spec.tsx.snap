// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/settings/settingreactiveinput should construct 1`] = `
<React.Fragment>
  <RenderToggleSetting
    checked={false}
    isReadOnly={false}
    onToggleInputChange={[Function]}
  />
</React.Fragment>
`;
